### PR TITLE
Custom ec2 ami, instance type and region

### PR DESCRIPTION
**Pull request by evansiroky**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

There currently isn't an issue for this, but here is what this PR does:

- allows customizing an ec2 deployment in the following ways:
  - setting a custom AMI ID for graph builds only
  - setting a custom instance type for graph builds only
  - setting a custom region for the deployment
- modifies the DeployJob to be able to do the following
  - do graph building with a separate instance type and/or AMI ID (this could be a substitute resolution for #265).
  - upload a graph build report if one was generated
- modifies other controllers to support custom regions
- Adds better debug information to deploy log statements
- skips request/response logging for job statuses

### CODE REVIEW

File: src/main/java/com/conveyal/datatools/manager/jobs/DeployJob.java [hunk 1]
```diff
@@ -621,19 +653,32 @@ private void replaceEC2Servers() {
...
-        String amiId = otpServer.ec2Info.amiId;
-        if (amiId == null) {
+        // Pick proper ami depending on whether graph is being built and what is defined.
+        String amiId;
+        if (!graphAlreadyBuilt && otpServer.ec2Info.buildAmiId != null) {
+            amiId = otpServer.ec2Info.buildAmiId;
+        } else if (otpServer.ec2Info.amiId != null) {
+            amiId = otpServer.ec2Info.amiId;
+        } else {
             amiId = DEFAULT_AMI_ID;
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/datatools/manager/jobs/DeployJob.java]: Should this be its own method?

File: src/main/java/com/conveyal/datatools/manager/jobs/DeployJob.java [hunk 2]
```diff
@@ -621,19 +653,32 @@ private void replaceEC2Servers() {
...
+        }
+        // Pick proper instance type depending on whether graph is being built and what is defined.
+        String instanceType;
+        if (!graphAlreadyBuilt && otpServer.ec2Info.buildInstanceType != null) {
+            instanceType = otpServer.ec2Info.buildInstanceType;
+        } else if (otpServer.ec2Info.instanceType != null) {
+            instanceType = otpServer.ec2Info.instanceType;
+        } else {
+            instanceType = DEFAULT_INSTANCE_TYPE;
         }
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/datatools/manager/jobs/DeployJob.java]: Should it be its own method?

File: src/main/java/com/conveyal/datatools/manager/jobs/MonitorServerStatusJob.java [hunk 3]
```diff
@@ -130,26 +133,27 @@ public void jobLogic() {
             wait("graph build/download check: " + graphStatusUrl);
             graphIsAvailable = checkForSuccessfulRequest(graphStatusUrl);
             if (jobHasTimedOut()) {
-                message = String.format("Job timed out while waiting for graph build/download (%s). If this was a graph building machine, it may have run out of memory.", instance.getInstanceId());
+                message = "Job timed out while waiting for graph build/download. If this was a graph building machine, it may have run out of memory.";
                 LOG.error(message);
-                status.fail(message);
+                failJob(message);
                 return;
```
Comment (binh-dam-ibigroup) [inline src/main/java/com/conveyal/datatools/manager/jobs/MonitorServerStatusJob.java]: Move `LOG.error(...)` inside method `failJob()` (3+ instances)

File: src/main/java/com/conveyal/datatools/manager/jobs/DeployJob.java [hunk 4]
```diff
@@ -80,8 +80,7 @@
     private static final Logger LOG = LoggerFactory.getLogger(DeployJob.class);
     private static final String bundlePrefix = "bundles";
     public static final String DEFAULT_INSTANCE_TYPE = "t2.medium";
-    private static final String AMI_CONFIG_PATH = "modules.deployment.ec2.default_ami";
-    private static final String DEFAULT_AMI_ID = DataManager.getConfigPropertyAsText(AMI_CONFIG_PATH);
+    public static final String AMI_CONFIG_PATH = "modules.deployment.ec2.default_ami";
```
Comment (landonreed) [inline src/main/java/com/conveyal/datatools/manager/jobs/DeployJob.java]: This constant is now only used in this file for a log statement. This makes me think that it should probably be moved to EC2Info?

File: src/main/java/com/conveyal/datatools/manager/models/EC2Info.java [hunk 5]
```diff
@@ -53,4 +57,36 @@ public boolean hasSeparateGraphBuildConfig() {
...
+     * @param graphAlreadyBuilt whether or not a graph has already been built. If false, this means a build ami should
+     *                          be used if available.
+     */
+    public String getAmiId(boolean graphAlreadyBuilt) {
+        if (!graphAlreadyBuilt && buildAmiId != null) {
+            return buildAmiId;
+        } else if (amiId != null) {
+            return amiId;
+        } else {
+            return DataManager.getConfigPropertyAsText(AMI_CONFIG_PATH);
```
Comment (landonreed) [inline src/main/java/com/conveyal/datatools/manager/models/EC2Info.java]: Maybe this final else clause should check whether the config value is missing and log an error if so?

File: src/main/java/com/conveyal/datatools/common/utils/SparkUtils.java [hunk 6]
```diff
@@ -234,6 +234,8 @@ public static void logRequestOrResponse(
             LOG.warn("Request object is null. Cannot log.");
             return;
         }
+        // don't log job status requests/responses, they clutter things up
+        if (request.pathInfo().contains("status/jobs")) return;
```
Comment (landonreed) [inline src/main/java/com/conveyal/datatools/common/utils/SparkUtils.java]: Lol, I feel like I've been advocating for this for a long time... what made you change your mind?

File: src/main/java/com/conveyal/datatools/manager/controllers/api/ServerController.java [hunk 7]
```diff
@@ -263,9 +284,23 @@ private static void validateFields(Request req, OtpServer server) throws HaltExc
             // If a server's ec2 info object is not null, it must pass a few validation checks on various fields related to
             // AWS. (e.g., target group ARN and instance type).
             if (server.ec2Info != null) {
+                // do some custom items if a custom region should be used
```
Comment (landonreed) [inline src/main/java/com/conveyal/datatools/manager/controllers/api/ServerController.java]: Don't love the first part of this comment.

File: src/main/java/com/conveyal/datatools/manager/models/EC2Info.java [hunk 8]
```diff
@@ -27,10 +31,62 @@ public EC2Info () {}
     public String securityGroupId;
     /** The Amazon machine image (AMI) to be used for the OTP EC2 machines. */
     public String amiId;
+    /**
+     * The AWS-style instance type (e.g., t2.medium) to use for new EC2 machines used specifically for graph building.
+     * Defaults to {@link com.conveyal.datatools.manager.models.EC2Info#instanceType} if null during deployment.
+     */
+    public String buildInstanceType;
+    /** The Amazon machine image (AMI) to be used for the OTP EC2 machine used specifically for graph building. */
+    public String buildAmiId;
```
Comment (landonreed) [inline src/main/java/com/conveyal/datatools/manager/models/EC2Info.java]: Could you also add default value here (like you have for `buildInstanceType`)?

File: src/main/java/com/conveyal/datatools/manager/models/EC2Info.java [hunk 9]
```diff
@@ -27,10 +31,62 @@ public EC2Info () {}
...
     /** The IAM instance profile ARN that the OTP EC2 server should assume. For example, arn:aws:iam::123456789012:instance-profile/otp-ec2-role */
     public String iamInstanceProfileArn;
     /** The AWS key file (.pem) that should be used to set up OTP EC2 servers (gives a way for admins to SSH into machine). */
     public String keyName;
     /** The target group to deploy new EC2 instances to. */
     public String targetGroupArn;
+    /** An optional custom AWS region */
+    public String region;
```
Comment (landonreed) [inline src/main/java/com/conveyal/datatools/manager/models/EC2Info.java]: Could you add an example value (e.g., `us-east-1`)?

File: src/main/java/com/conveyal/datatools/manager/jobs/DeployJob.java [hunk 10]
```diff
@@ -832,20 +867,31 @@ private String constructUserData(boolean graphAlreadyBuilt) {
...
         }
         // Determine if graph build/download was successful (and that Graph.obj is not zero bytes).
         lines.add(String.format("FILESIZE=$(wc -c <%s)", graphPath));
         lines.add(String.format("[ -f %s ] && (($FILESIZE > 0)) && GRAPH_STATUS='SUCCESS' || GRAPH_STATUS='FAILURE'", graphPath));
+        // Re-upload user data log before indicating that graph build/download is complete.
+        lines.add(uploadUserDataLogCommand);
         // Create file with bundle status in web dir to notify Data Tools that download is complete.
         lines.add(String.format("sudo echo $GRAPH_STATUS > $WEB_DIR/%s", GRAPH_STATUS_FILE));
-        // Re-upload user data log before final command (and before optional shutdown statement).
-        lines.add(uploadUserDataLogCommand);
```
Comment (landonreed) [inline src/main/java/com/conveyal/datatools/manager/jobs/DeployJob.java]: I don't quite understand why this line was moved up.

Comment (evansiroky) [reply src/main/java/com/conveyal/datatools/manager/jobs/DeployJob.java]: I wanted to make sure that the logs were uploaded before echoing the success into the graph status file. If the DeployJob detects that the graph building is done and terminates the instance before the logs can get uploaded that's not good.

File: src/main/java/com/conveyal/datatools/manager/models/EC2Info.java (hunk 5 above)
Comment (evansiroky) [reply src/main/java/com/conveyal/datatools/manager/models/EC2Info.java]: There is already a check for null in the DeployJob.

**Comment by evansiroky** [comment]
See latest changes.

**Pull request merged**